client/login/controller.js

```diff
@@ -159,3 +157,37 @@ export function redirectJetpack( context, next ) {
 	}
 	next();
 }
+
+export function setShouldSSRLogin( context, next ) {
+	if ( ! context.SSRValidators ) {
+		next(); // practically somehow the initSSRValidators was not initialized, so adding more is not necessary
+		return;
+	}
+
+	context.SSRValidators = Object.assign( context.SSRValidators, {
+		// Overrides the query validator, which by default would ask no parameters
+		query: () => {
+			const whitelistedQueryKeys = [ 'client_id', 'signup_flow', 'redirect_to' ];
+			const queryKeys = Object.keys( context.query );
+
+			// if there are any parameters, they must be ONLY the ones in the whitelist
+			return queryKeys.length === intersection( queryKeys, whitelistedQueryKeys ).length;
+		},
+
+		// When it exists, the redirect_to query parameter must satisfy a specific format
+		query__redirect_to: () => {
+			if ( ! context.query.redirect_to ) {
+				return true; // nothing to check
+			}
+
+			const redirectToDecoded = decodeURI( context.query.redirect_to );
```

Comment:
Did you already consider using `decodeURIComponent()` here?
I think it might be the better choice, but maybe there's a reason you chose not to use it.
Ref: https://stackoverflow.com/questions/747641/what-is-the-difference-between-decodeuricomponent-and-decodeuri